Add new Zehnder ComfoAir 350 integration

mtdcr opened the pull request:
## Description:
This integrates support for Zehnder ComfoAir 350 ventilation units and compatibles. This is not to be confused with the comfoconnect component, which integrates the newer Q350 units.

It allows reading temperature sensors and setting fan speeds. Currently it expects a CC Ease controller to be present.

Documentation will be written if feedback indicates a chance to merge.

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io):** home-assistant/home-assistant.io#10964

## Example entry for `configuration.yaml`:
```yaml
comfoair:
  name: ComfoAir
  serial_port: /dev/ttyS0
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [x] Untested files have been added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

mtdcr committed:
Add new Zehnder ComfoAir 350 integration

mtdcr commented:
@frenck: I've added docs now.

MartinHjelmare commented:
If this is the same brand as the  comfoconnect integration, just a different model, we should combine both models into a single integration. We can use a config option to select between the models and logic paths.

It's ok to depend on different libraries for the different models.

MartinHjelmare commented:
CC @michaelarnauts

michaelarnauts commented:
The two models are quite different, but are from the same vendor. Comfoconnect is using a bridge to communicate, and this uses a serial connection. I don't think there is code to share between the two implementations, so I'm not sure if there is a benefit to merge them.

mtdcr commented:
Yes, they really don't have anything in common.

MartinHjelmare commented:
It's our standard to merge different models of devices from the same brand into the same integration.

michaelarnauts commented:
We could just merge the files and use different classes, but the register for the sensors are also different, so the config validation will probably be quite messy.

mtdcr commented:
We can also rename comfoair to jsstorkair to use a different but compatible brand.

In any case, merging this code into a module called comfoconnect seems insane, because comfoconnect is just a bridge you can buy as an accessory to connect newer, incompatible devices to a network.

michaelarnauts commented:
@MartinHjelmare I can't find a requirement that components from a vendor have to be merged together. I can understand this from a user perspective, but it just doesn't make sense to merge them since they are completely different implementations for different devices with different features.

I took a quick look at the integrations list and found a bunch of other examples:

Both Aruba Presence Detection, but different implementations:
* https://www.home-assistant.io/integrations/aruba/
* https://www.home-assistant.io/integrations/cppm_tracker/

Both Bluetooth trackers, but a different version:
* https://www.home-assistant.io/integrations/bluetooth_le_tracker/
* https://www.home-assistant.io/integrations/bluetooth_tracker/

Both Cisco Presence Detection, but different implementations:
* https://www.home-assistant.io/integrations/cisco_ios/
* https://www.home-assistant.io/integrations/cisco_mobility_express/

All three components are Media Players of Denon:
* https://www.home-assistant.io/integrations/denonavr/
* https://www.home-assistant.io/integrations/denon/
* https://www.home-assistant.io/integrations/heos/

I've stopped looking at the letter D, I'm sure there are a lot more examples.

balloob commented:
Examples in our code base cannot be used as argument why we should do something. The rules evolve over time and there will always be code not following our latest set of rules.

I think that in this case, it's ok to keep them separate, since they are a complete different API (serial vs IP)

michaelarnauts commented on homeassistant/components/comfoair/__init__.py:
```diff
@@ -0,0 +1,119 @@
+
+    def __init__(self, hass: HomeAssistantType, config: ConfigType):
+        """Initialize the ComfoAir component."""
+        from comfoair.asyncio import ComfoAir
```
Import should be moved to the top. See #27284

michaelarnauts commented on homeassistant/components/comfoair/sensor.py:
```diff
@@ -0,0 +1,158 @@
+        _LOGGER.debug("Sensor type: %s", sensor_type)
+
+        if sensor_type not in SENSOR_TYPES:
+            _LOGGER.warning("Sensor type: %s is not a valid sensor.", sensor_type)
```
Logging messages should not end with a period

michaelarnauts replied:
Actually, you should use vol to validate the sensors config. I recently did the same for comfoconnect, see https://github.com/home-assistant/home-assistant/commit/3d5b007c6bc63be50a4aa5bb8e45e9366bdbd201

mtdcr replied:
I'll remove the data selector as recommended by Martin in https://github.com/home-assistant/home-assistant/pull/27961#discussion_r350895455

michaelarnauts commented on homeassistant/components/comfoair/fan.py:
```diff
@@ -0,0 +1,116 @@
+        return self.async_turn_off()
+
+    def __ca_set_speed(self, speed):
+        _LOGGER.debug("Changing fan speed to %s.", SPEED_MAPPING[speed])
```
Remove the period at the end.

mtdcr commented:
@michaelarnauts: I addressed your comments and rebased my branch. However, your recommendation to use vol caused some trouble, i.e. `CONF_RESOURCES` did not appear in `config`. That's why I left it out for now. You can see the additional patch here: https://github.com/mtdcr/home-assistant/commit/5846640fbdb8b4e777aa9af8b6ba47a27d721af1
Can you please take a brief look at it? Maybe you can see at a glance what's missing to properly glue it together.

I also removed the dependency on pyserial-asyncio, which was only indirect, and moved another import to the top.

It would be nice if someone with the necessary privileges could remove the „docs-missing“ label.

michaelarnauts commented:
> @michaelarnauts: I addressed your comments and rebased my branch. However, your recommendation to use vol caused some trouble, i.e. `CONF_RESOURCES` did not appear in `config`. That's why I left it out for now. You can see the additional patch here: [mtdcr@5846640](https://github.com/mtdcr/home-assistant/commit/5846640fbdb8b4e777aa9af8b6ba47a27d721af1)
> Can you please take a brief look at it? Maybe you can see at a glance what's missing to properly glue it together.
> 
> I also removed the dependency on pyserial-asyncio, which was only indirect, and moved another import to the top.
> 
> It would be nice if someone with the necessary privileges could remove the „docs-missing“ label.

I don't immediately see why the PLATFORM_SCHEMA is wrong. Maybe @MartinHjelmare can see more?

You can fix your import errors by running `isort` on your files.

MartinHjelmare commented on homeassistant/components/comfoair/__init__.py:
```diff
@@ -0,0 +1,118 @@
+"""Support to control a Zehnder ComfoAir 350 ventilation unit."""
+
+import logging
+import voluptuous as vol
```
Please add a blank line between standard library and 3rd party import groups.

MartinHjelmare commented on homeassistant/components/comfoair/__init__.py:
```diff
@@ -0,0 +1,118 @@
+from homeassistant.helpers.dispatcher import async_dispatcher_send
+from homeassistant.helpers.typing import ConfigType, HomeAssistantType
+import homeassistant.helpers.config_validation as cv
+from comfoair.asyncio import ComfoAir
```
Please group this together with `voluptuous` as a 3rd party import.

MartinHjelmare commented on homeassistant/components/comfoair/__init__.py:
```diff
@@ -0,0 +1,118 @@
+        hass.async_create_task(coro)
+
+    async def _ca_service_virtualkey(service):
+        ev_mask = service.data.get(ATTR_MASK)
```
Please use `dict[key]` for required schema keys and keys with default schema values.

MartinHjelmare commented on homeassistant/components/comfoair/__init__.py:
```diff
@@ -0,0 +1,118 @@
+        coro = discovery.async_load_platform(hass, com, DOMAIN, {}, config)
+        hass.async_create_task(coro)
+
+    async def _ca_service_virtualkey(service):
```
We don't need to prefix this function with underscore.

MartinHjelmare commented on homeassistant/components/comfoair/__init__.py:
```diff
@@ -0,0 +1,118 @@
+        self._cache = {}
+
+        self._ca = ComfoAir(config[DOMAIN].get(CONF_SERIAL_PORT))
+        self._ca.add_listener(self._event)
```
Does this have a side effect? We try to avoid side effects in init method.

mtdcr replied:
It just adds the callback to a list. The callback won't get executed until the mainloop starts reading data from the serial port.

MartinHjelmare commented on homeassistant/components/comfoair/__init__.py:
```diff
@@ -0,0 +1,118 @@
+        """Connect to a serial port or socket."""
+        await self._ca.connect(self._hass.loop)
+
+    async def stop(self):
```
We should probably schedule this on home assistant stop event.

MartinHjelmare commented on homeassistant/components/comfoair/fan.py:
```diff
@@ -0,0 +1,116 @@
+class ComfoAirFan(FanEntity):
+    """Representation of the ComfoAir fan platform."""
+
+    def __init__(self, hass, ca: ComfoAirModule) -> None:
```
Don't pass in `hass`. It will be set on the entity when it has been added to home assistant.

MartinHjelmare commented on homeassistant/components/comfoair/fan.py:
```diff
@@ -0,0 +1,116 @@
+            _update_state(self._sensor_id, data)
+
+        # Register for dispatcher updates
+        async_dispatcher_connect(
```
Please move this to `async_added_to_hass` entity coroutine method.

MartinHjelmare commented on homeassistant/components/comfoair/fan.py:
```diff
@@ -0,0 +1,116 @@
+    @property
+    def speed_list(self):
+        """List of available fan modes."""
+        return [SPEED_OFF, SPEED_LOW, SPEED_MEDIUM, SPEED_HIGH]
```
This constant list can be moved to a module level constant and returned here.

MartinHjelmare commented on homeassistant/components/comfoair/fan.py:
```diff
@@ -0,0 +1,116 @@
+        # shouldn't happen
+        return self.async_turn_off()
+
+    def __ca_set_speed(self, speed):
```
Why do we prefix with double underscore?

MartinHjelmare commented on homeassistant/components/comfoair/sensor.py:
```diff
@@ -0,0 +1,157 @@
+        "Comfort Temperature",
+        TEMP_CELSIUS,
+        "mdi:thermometer",
+        0xD2,
```
This is protocol specific information that doesn't belong in home assistant. Please abstract that into the device library.

MartinHjelmare commented on homeassistant/components/comfoair/sensor.py:
```diff
@@ -0,0 +1,157 @@
+    unit = hass.data[DOMAIN]
+
+    sensors = []
+    for resource in config.get(CONF_RESOURCES, []):
```
We want to avoid having config option for entity types if possible. See:
https://github.com/home-assistant/architecture/blob/master/adr/0003-monitor-condition-and-data-selectors.md

MartinHjelmare commented on homeassistant/components/comfoair/sensor.py:
```diff
@@ -0,0 +1,157 @@
+        sensors.append(
+            ComfoAirSensor(
+                hass,
+                name="%s %s" % (unit.name, SENSOR_TYPES[sensor_type][0]),
```
Please use f-strings instead of old style string formatting.

MartinHjelmare commented on homeassistant/components/comfoair/sensor.py:
```diff
@@ -0,0 +1,157 @@
+
+        def _update_state(data):
+            bits = BitArray(data)
+            value = bits[self._offset : self._offset + self._size].uint
```
We shouldn't need to handle bits and offsets in home assistant. Please move that to the library.

mtdcr committed:
comfoair: Address Martin's comments as far as pycomfoair 0.0.1 permits

MartinHjelmare commented on homeassistant/components/comfoair/fan.py:
```diff
@@ -87,30 +90,30 @@ def speed(self):
     def speed_list(self):
         """List of available fan modes."""
-        return [SPEED_OFF, SPEED_LOW, SPEED_MEDIUM, SPEED_HIGH]
+        return list(SPEED_MAPPING.values())
```
Please create a new constant at the module level so we don't need to do an extra call on every access. It's ok to depend on the `SPEED_MAPPING` constant when creating the other constant.

MartinHjelmare commented on homeassistant/components/comfoair/sensor.py:
```diff
@@ -0,0 +1,154 @@
+    unit = hass.data[DOMAIN]
+
+    sensors = []
+    for resource in SENSOR_TYPES.keys():
```
Iterating a dict, iterates the keys by default.
```suggestion
    for resource in SENSOR_TYPES:
```

mtdcr commented:
@MartinHjelmare, thanks for your thorough review! As the remaining two changes require an updated library, would it be possible to handle them in a follow-up PR? It will make the library more complex and therefore will need some time to develop and test. The current version works pretty stable, AFAICT.

MartinHjelmare commented:
I think we should wait here for an updated library. We don't want to add protocol specific code in home assistant. It's one of our basic rules.

balloob commented:
PR has gone stale. Closing it.
